[aggregator] implement time sampler sharding in the `Demultiplexer`.

remeh opened the pull request:
### What does this PR do?

I've introduced the `Demultiplexer` in #9657, which is intended to provide a better API to all data transmitted the Datadog Agent pipeline, but which will also be helpful to improve the CPU usage when more cores are available by better distributing the work on different routines.

First point has been implemented by the introduction of the `AgentDemultiplexer` implementation available since the previous pull-request.

Second point is starting to be addressed in this pull-request, by introducing a change in the `TimeSampler` and by sharding multiple of them in the `Demultiplexer`.

~This PR is also making the `TimeSampler`s flushing directly into a `Serializer`, instead of the `BufferedAggregator` being responsible of this.~
The `Demultiplexer` also becomes responsible of triggering the flush of all samplers + the `BufferedAggregator` and to send all this data to the serializer.

### Motivation

* Offer better performance in the DogStatsD pipeline when it is possible to use multiple cores (i.e. diminish the drop by using more CPU)
* The `BufferedAggregator` is still responsible of the sampling of the check samples but not of the time samples (i.e. DogStatsD), which is reducing its scope already, and which will continue to be reduced once the different misc parts (e.g. event platform forwarding, orchestrator data forwarder) is properly attached to the `Demultiplexer`.

### Additional Notes

1. The shard key used to distribute the metric sample on the time samplers is using 32 bits of the generated context key + the [fastrange](https://github.com/lemire/fastrange) which should offer a proper distribution on time samplers.
2. ~Streaming of series to the serializer has been introduced more or less at the same time that this branch has been developed, a following pull-request will remove the logic duplication used to stream the series from the `TimeSampler` into its `Serializer`.~ - Done in #10890 
3. The Serverless Agent doesn't need a `BufferedAggregator` anymore since it is only sampling time samples. However, the `BufferedAggregator` is still here due to some dependencies on the (unused) service checks/events support, I will remove this in a following pull-request.
4. A `tagsBuffer` is constantly reseted and reused with the tags received with a time sample to compute its shard key. This `tagsBuffer` equivalent and its corresponding hash is recreated later while generating the final `ContextKey` of the time sample, it would be a performance improvement to keep the one from the shard key and to just re-use it while computing the `ContextKey`.

### Todos left before merging

  * [x] Some telemetry has not been properly restored yet.
  * [x] Document how to QA this pull-request.
  * [x] A couple unit tests seems to still be broken
  * [x] Write a release-note entry.

### Possible Drawbacks / Trade-offs

Using multiple pipelines on host not dedicating multiple cores to the Datadog Agent could lead to worst performance (with the runtime scheduler having to schedule too many goroutines).
Note that the Agent is defaulting on only one pipeline for now, and with this default, it should offer more or less the same performance as it was before this pull-request.

### Describe how to test/QA your changes

* Runs the Agent
* Sends DogStatsD metrics using a DogStatsD client
* Validate that the metric are reaching a dashboard with proper values

Multiple scenarios are interesting to QA here, for this scenario, it will be interesting to have some code around using a DogStatsD client to generate various metrics. Let's say we have this Ruby script, which is generating 30 different metrics.

```ruby
require '../lib/datadog/statsd'
require 'logger'

statsd = Datadog::Statsd.new('localhost', 8125, logger: Logger.new(STDOUT))

200.times do |i| # multiple sends
    15.times do |j| # generate different metrics
        statsd.increment('metric_one.test', tags: ['team:agent-core', 'scenario:qa', 'environment:dev', "tag:#{j}"])
        statsd.increment('metric_two.test', tags: ['team:agent-core', 'scenario:qa', 'environment:dev', "tag:#{j}"])
    end
    statsd.flush(sync: true)
    sleep(0.2)
end

puts("Done.")
```

## Main scenario: one pipeline

Run the Datadog Agent with default settings and validate that all the 30 metrics reach the intake properly. To do that easily, use the Metrics Explorer to display "One graph per" and use the tag `tag` to explode in multiple graphs. All metrics should be there and their values should be consistent between each others.

## New scenario: multiple pipelines (manually configured)

Set the `dogstatsd_pipeline_count` to `2` or `3` in your configuration, then run the "Main scenario".

## New scenario: multiple pipelines (auto-adjust)

You can run the QA scenario described here: https://github.com/DataDog/datadog-agent/pull/10944

If you have your Datadog Agent running on a host with many cores (meaning it will startup with `GOMAXPROCS > 6` for instance). Set the configuration flag `dogstatsd_pipeline_autoadjust` to `true`, validate it is auto-adjusting the amount of pipelines (see #10944 logs messages in QA section) and run the Main scenario.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

remeh committed:
[aggregator] implement time sampler sharding in the `Demultiplexer`.

ogaca-dd commented on pkg/dogstatsd/batch.go:
```diff
@@ -9,44 +9,95 @@ import (
+		b.tagsBuffer.Append(sample.Tags...)
+		h := b.keyGenerator.Generate(sample.Name, sample.Host, b.tagsBuffer)
+		b.tagsBuffer.Reset()
+		shardKey = fastrange(h, b.pipelineCount)
```
Can a round robin algorithm be used instead?

remeh replied:
No, it would lead to incorrect sampling of the data.

ogaca-dd commented on pkg/serverless/metrics/enhanced_metrics.go:
```diff
@@ -48,37 +50,37 @@ func getOutOfMemorySubstrings() []string {
 			SampleRate: 1,
-			Timestamp:  float64(end.UnixNano()),
-		}}
+			Timestamp:  float64(end.UnixNano()) / float64(time.Second),
```
Was there a bug before your change here?

remeh replied:
No bug here before, the Serverless Agent was sending it nano and / second later in the process, I've normalized to seconds everywhere.

ogaca-dd commented on pkg/serverless/metrics/enhanced_metrics_test.go:
```diff
@@ -9,154 +9,160 @@ import (
+	generatedMetrics := demux.WaitForSamples(100 * time.Millisecond)
 
-	assert.Equal(t, generatedMetrics, []metrics.MetricSample{{
+	assert.Equal(t, generatedMetrics[:1], []metrics.MetricSample{{
```
It would be nice to check if `generatedMetrics` is not empty

remeh replied:
I think this should not have been committed, I'll look into this.

ogaca-dd reviewed (commented):
Great work.
Note: I did a high level review of the PR.

remeh commented on pkg/aggregator/time_sampler.go:
```diff
@@ -20,6 +23,9 @@ type SerieSignature struct {
+
 // TimeSampler aggregates metrics by buckets of 'interval' seconds
 type TimeSampler struct {
 	interval                    int64
```
While I agree it has grew now that it is responsible of its own flushing, the `TimeSampler` has always be responsible of creating the sketches and series since it is... sampling them in time, so it is part of its responsibility.
As mention in the code though, the added code introduced to serialize the "stream series" since you introduced your change will have to be refactored with the Aggregator / other samplers at some point, but no reason to do it now afaik, especially if we want to keep this pull-request reviewable.

remeh replied:
Created the `timeSamplerWorker` to address this.

remeh commented on pkg/aggregator/aggregator_test.go:
```diff
@@ -499,31 +509,34 @@ func TestTags(t *testing.T) {
 
+	pc := config.Datadog.GetInt("dogstatsd_pipeline_count")
+	config.Datadog.Set("dogstatsd_pipeline_count", 1)
+	defer func() {
```
Just because sometimes, doing this in unit tests, I have to do it for multiple configuration fields, but you are right that I could not have used a lambda here that would make no difference.

remeh commented on pkg/aggregator/demultiplexer.go:
```diff
@@ -290,21 +366,64 @@ func (d *AgentDemultiplexer) Stop(flush bool) {
-func (d *AgentDemultiplexer) AddTimeSamples(samples []metrics.MetricSample) {
-	panic("not implemented yet.")
+// This function is called by different goroutines and is thread-safe.
+func (d *AgentDemultiplexer) AddTimeSamples(shard uint32, samples []metrics.MetricSample) {
```
Using a round-robin algorithm to distribute the metrics would lead to completely wrong data.

remeh commented on pkg/aggregator/demultiplexer.go:
```diff
@@ -330,27 +449,51 @@ func (d *AgentDemultiplexer) Aggregator() *BufferedAggregator {
 // ------------------------------
 
 // ServerlessDemultiplexer is a simple demultiplexer used by the serverless flavor of the Agent
 type ServerlessDemultiplexer struct {
```
You are right, I will do it in a following pull-request (doing it in this pull-request would create an awfully hard to read pull-request).

remeh commented on pkg/aggregator/demultiplexer.go:
```diff
@@ -382,20 +525,40 @@ func (d *ServerlessDemultiplexer) Stop(flush bool) {
+	// but these slices have a fixed size, we may have to grow some of them while using them.
+	ms := d.metricSamplePool.GetBatch()
+	for i := 0; i < len(samples); i++ {
+		if i < MetricSamplePoolBatchSize {
```
You are right, I'll change this.

remeh committed:
Merge remote-tracking branch 'origin/main' into remeh/sharded-aggregator-sharding

remeh committed:
[demultiplexer] use the same flush frequency in the `TimeSampler` as before.

vickenty commented on go.mod:
```diff
@@ -91,7 +91,7 @@ require (
 	github.com/alecthomas/repr v0.0.0-20181024024818-d37bc2a10ba1
 	github.com/avast/retry-go v3.0.0+incompatible
-	github.com/aws/aws-sdk-go v1.42.20
+	github.com/aws/aws-sdk-go v1.42.2
```
Looks like this will undo a package upgrade on main. Is this intentional?

vickenty commented on pkg/aggregator/demultiplexer.go:
```diff
@@ -199,11 +248,45 @@ func initAgentDemultiplexer(options DemultiplexerOptions, hostname string) *Agen
+// to be sent on the next flush.
+func (d *AgentDemultiplexer) AddAgentStartupTelemetry(agentVersion string) {
+	if agentVersion != "" {
+		d.AddTimeSamples(0, []metrics.MetricSample{{
```
Will the new slice end up in the `MetricSamplePool`?

vickenty commented on pkg/aggregator/time_sampler.go:
```diff
@@ -75,11 +134,121 @@ func (s *TimeSampler) addSample(metricSample *metrics.MetricSample, timestamp fl
+		case <-s.stopChan:
+			return
+		case ms := <-s.samples:
+			// do this telemetry here and not in the samplers goroutines, this way
```
But this is in the sampler goroutine though?

remeh replied:
Yup and effectively not much I can do about this actually, however, they did not show up in benchmarks so we are fine, I'm just removing the comment.

vickenty commented on pkg/aggregator/time_sampler.go:
```diff
@@ -75,11 +134,121 @@ func (s *TimeSampler) addSample(metricSample *metrics.MetricSample, timestamp fl
+		log.Debugf("Time Sampler #%d Flushing series to the forwarder", s.id)
+		var series metrics.Series
+		sketches := s.flush(float64(t.Unix()), &series) // XXX(remy): is this conversation correct? note that it is in second
+		// XXX(remy): better error management
```
Is this something you wanted to address in this PR?

remeh replied:
Done!

vickenty commented on pkg/aggregator/time_sampler.go:
```diff
@@ -75,11 +134,121 @@ func (s *TimeSampler) addSample(metricSample *metrics.MetricSample, timestamp fl
+		if s.serializer != nil {
+			//
+			// TODO(remy): restore all the telemetry
+			//
```
This too?

remeh replied:
Done!

vickenty commented on pkg/aggregator/time_sampler.go:
```diff
@@ -75,11 +134,121 @@ func (s *TimeSampler) addSample(metricSample *metrics.MetricSample, timestamp fl
+	} else {
+		log.Debugf("Time Sampler #%d Flushing series to the forwarder", s.id)
+		var series metrics.Series
+		sketches := s.flush(float64(t.Unix()), &series) // XXX(remy): is this conversation correct? note that it is in second
```
This looks correct, since we truncate it later anyway, though personally I'd keep the original `UnixNano()/time.Second` here, or just convert everything to use int64. Floating point timestamps look like they should be subsecond precision.

remeh committed:
[dogstatsd] workers count is function of how many aggregation pipeline are running.

remeh committed:
Merge remote-tracking branch 'origin/main' into remeh/sharded-aggregator-sharding

remeh committed:
[modules] restore `go.mod` with `main`.

remeh commented on pkg/serverless/metrics/enhanced_metrics.go:
```diff
@@ -48,37 +50,37 @@ func getOutOfMemorySubstrings() []string {
 	} else {
 		duration := end.Sub(start).Milliseconds()
-		metricsChan <- []metrics.MetricSample{{
+		demux.AddTimeSamples(0, []metrics.MetricSample{{
```
I agree, I had a first draft with functions not exposing sharding. I'll give it a second try.

remeh commented on pkg/aggregator/time_sampler.go:
```diff
@@ -53,11 +101,22 @@ func (s *TimeSampler) isBucketStillOpen(bucketStartTimestamp, timestamp int64) b
 
 // Add the metricSample to the correct bucket
-func (s *TimeSampler) addSample(metricSample *metrics.MetricSample, timestamp float64) {
+func (s *TimeSampler) addSamples(samples []metrics.MetricSample) {
```
That's a good idea.

remeh replied:
This lead to a fairly big change, do you mind if l address this a in a follow-up PR?

remeh committed:
[aggregator] add `GetDogStatsDPipelinescount()` in the `Demultiplexer` and address a couple feedback

Also fix a merge mistake in the serverless agent.

remeh committed:
[demultiplexer] better `AddTimeSample`/`AddTimeSampleBatch` API for the Demultiplexer.

It is also introducing the `MetricSampleBatch` type.

remeh committed:
[time_sampler] extracted the process loop (flush, samples receiving) in a `timeSamplerWorker`.

Introduced a `Stop()` and `Flush()` method in the `TimeSampler`.

remeh commented on pkg/aggregator/time_sampler.go:
```diff
@@ -75,11 +132,121 @@ func (s *TimeSampler) addSample(metricSample *metrics.MetricSample, timestamp fl
+
+	if waitForSerializer {
+		<-done
+	}
```
That's how you did it in the `pkg/aggregator`, you're sure I should change it to be after?

ogaca-dd replied:
Indeed what you did is correct.

remeh committed:
[demultiplexer] proper telemetry and error management in the timeSamplerWorker.

remeh committed:
[time_sampler] the `flushInterval` could be 0 when no automatic flush should run.

maxday commented on pkg/aggregator/demultiplexer.go:
```diff
@@ -369,8 +521,8 @@ func (d *ServerlessDemultiplexer) Run() {
 	}
 	log.Debug("Forwarder started")
 
+	log.Info("Demultiplexer started")
```
Could we keep this as Debug?

maxday commented on pkg/serverless/metrics/enhanced_metrics.go:
```diff
@@ -128,39 +130,42 @@ func GenerateEnhancedMetricsFromReportLog(initDurationMs float64, durationMs flo
-func SendOutOfMemoryEnhancedMetric(tags []string, time time.Time, metricsChan chan []metrics.MetricSample) {
-	incrementEnhancedMetric(OutOfMemoryMetric, tags, float64(time.UnixNano()), metricsChan)
+func SendOutOfMemoryEnhancedMetric(tags []string, t time.Time, demux aggregator.Demultiplexer) {
+	incrementEnhancedMetric(OutOfMemoryMetric, tags, float64(t.UnixNano())/float64(time.Second), demux)
```
just to make sure, are we changing the unit of these metrics?

maxday commented on pkg/dogstatsd/server.go:
```diff
@@ -446,10 +455,12 @@ func (s *Server) ServerlessFlush() {
-	agg.ServerlessFlush <- true
-	<-agg.ServerlessFlushDone
+	// We add 10 seconds to the interval to ensure that we're getting the whole sketches bucket
+	s.demultiplexer.ForceFlushToSerializer(start.Add(time.Second*10), true)
```
We used ServerlessFlushDone to make sure the flush was synchronous, is `ForceFlushToSerializer` blocking?

remeh replied:
Yep I've taken care of this:

Previous to my change, this `ServerlessFlushDone` was filled [here in the aggregator](https://github.com/DataDog/datadog-agent/pull/10532/files#diff-0940f1dc68973941178797f388ffb31d3318abeefa87b7f0b66d575915a0e575L834), it was already using `waitForSerializer` true so I _think_ it was actually redundant. 

With the change in this pull-request, setting `waitForSerializer` parameter to `true` let the flushing mechanism that we are waiting on a channel for the end of the flush: 
  1. here [we trigger a flush](https://github.com/DataDog/datadog-agent/pull/10532/files#diff-29b35a8b2b30e7c024255d7c86cb52e8b6f7b5401f441c7ad005a9211ec4dc39R68-R76) and you can see that this `flushTrigger` is accompanied with a channel 
  2. this channel will receive something only once the flush is over, [here](https://github.com/DataDog/datadog-agent/pull/10532/files#diff-1bc4614e41a3b27abc34f3f211edc7e59151a6180117f55d07de8f4b99812a40R102-R106)
  3. and we are correctly [waiting for this message](https://github.com/DataDog/datadog-agent/pull/10532/files#diff-29b35a8b2b30e7c024255d7c86cb52e8b6f7b5401f441c7ad005a9211ec4dc39R78-R79)

remeh committed:
[demultiplexer] log level of the starting logs in the ServerlessDemultiplexer

remeh committed:
Merge remote-tracking branch 'origin/main' into remeh/sharded-aggregator-sharding

remeh committed:
[modules] fix `go.sum` after the merge of main.

vickenty commented on pkg/aggregator/time_sampler.go:
```diff
@@ -29,38 +29,17 @@ type TimeSamplerID int
-type FlushCommand struct {
-	Time      time.Time
-	BlockChan chan struct{}
+	worker *timeSamplerWorker
```
Would it make sense to remove this circular dependency (sampler and worker have pointers to each other)? If I understand correctly, Demultiplexer only interacts with the worker (via channels), so perhaps it would make sense to store worker pointers in the demultiplexer, leaving the sampler only as a field of the worker.

maxday reviewed (approved):
💯 LGTM for the serverless part! Thanks

remeh committed:
[demultiplexer] the process loop + flush mechanism is started independently of the TimeSampler

It is also tearing down the samplers properly in the Stop() method of the Demultiplexer,
which was missing.

remeh committed:
[time_sampler] cosmetic: functions instance acronym.

vickenty commented on pkg/aggregator/time_sampler_test.go:
```diff
@@ -30,23 +30,24 @@ func generateSerieContextKey(serie *metrics.Serie) ckey.ContextKey {
+	sampler := NewTimeSampler(TimeSamplerID(0), 10, tags.NewStore(false, "test"))
+	worker := newTimeSamplerWorker(sampler, 15*time.Second, 100, metrics.NewMetricSamplePool(32), &serializer.MockSerializer{}, flushAndSerializeInParallel{enabled: false})
+	go worker.run()
+	return sampler, worker
```
It looks like the tests do not interact with the worker apart from `defer worker.stop()`. Would it make sense to remove it altogether?

remeh committed:
[demultiplexer] remove unecessary `statsd.samplers` + don't run the worker in the TimeSampler tests.

vickenty reviewed (approved):
:rocket:

remeh committed:
Merge remote-tracking branch 'origin/main' into remeh/sharded-aggregator-sharding

remeh committed:
[modules] go mod tidy.

ogaca-dd commented on pkg/aggregator/demultiplexer.go:
```diff
@@ -285,33 +386,81 @@ func (d *AgentDemultiplexer) Stop(flush bool) {
 	panic("not implemented yet.")
 }
 
+// GetDogStatsDPipelinesCount returns how many sampling pipeline are running for
```
```suggestion
// GetDogStatsDPipelinesCount returns how many sampling pipelines are running for
```

ogaca-dd commented on pkg/aggregator/time_sampler.go:
```diff
@@ -52,12 +65,18 @@ func (s *TimeSampler) isBucketStillOpen(bucketStartTimestamp, timestamp int64) b
 	contextKey := s.contextResolver.trackContext(metricSample, timestamp)
 	bucketStart := s.calculateBucketStart(timestamp)
 
+	// log.Infof("TimeSampler #%d processed sample '%s' tags '%s': %s", s.id, metricSample.Name, metricSample.Tags)
```
To remove?

ogaca-dd reviewed (approved):
LGTM.
I did not spot any issues but I didn't compare how the new code behaves versus the previous one. I let you decide if you think another review is needed or not.

remeh commented on pkg/serverless/metrics/enhanced_metrics.go:
```diff
@@ -131,7 +131,9 @@ func GenerateEnhancedMetricsFromReportLog(initDurationMs float64, durationMs flo
 
-	demux.AddTimeSamples(0, enhancedMetrics)
+	for _, metric := range enhancedMetrics {
+		demux.AddTimeSample(metric)
```
It is better to not add the "batch" complexity here:
  * no real needs for performances so no big deal using `AddTimeSample`
  * easier API call (no sharding appearance)
  * the MetricSampleBatch has to be retrieved filled and sent, so again easier API call this way.

remeh commented on pkg/aggregator/demultiplexer.go:
```diff
@@ -527,7 +558,11 @@ func (d *ServerlessDemultiplexer) Run() {
 // Stop stops the wrapped aggregator and the forwarder.
 func (d *ServerlessDemultiplexer) Stop(flush bool) {
-	d.aggregator.Stop(flush)
+	d.statsdWorker.flush(time.Now(), flush == true)
```
Fair point

remeh committed:
[demultiplexer] address a few feedback on comments.

remeh committed:
[demultiplexer] on Stop, do not use the public ForceFlushToSerializer which is locking.

Both these methods use the mutex to be thread-safe, one calling the other is blocking,
I've extracted the logic part of ForceFlushToSerializer into a private one not applying
the lock.

remeh committed:
[time_sampler_worker] SendSketches only if there is at least one sketch to serialize.

remeh committed:
[tests] this test is only sending series, not sketches, we should not receive a `SendSketch` call.

remeh committed:
Merge remote-tracking branch 'origin/main' into remeh/sharded-aggregator-sharding

remeh committed:
[general] add release note entry.

remeh committed:
[demultiplexer] the `Demultiplexer` is responsible of flushing all samplers/service checks/events to the serializer (#10890)

* [demultiplexer] the `Demultiplexer` is responsible of flushing all samplers.

* [demultiplexer] address review feedback

  * use a different type for simple triggers and for flush triggers
  * telemetry MainFlushTime and flush count fixed
  * restore `log_payloads` debug feature

* [demultiplexer] fix npe when the flush interal is 0

* [demultiplexer] notify as soon as series and sketches instead of waiting for other parts.

remeh committed:
Merge branch 'main' into remeh/sharded-aggregator-sharding

remeh committed:
[demultiplexer] fix an error done during the merge conflict resolution.

remeh committed:
[tests] avoid the flakiness of this test.

Not the best way but we're speaking ms executions here.
